src/hotspot/share/opto/subnode.cpp

```diff
@@ -1218,6 +1213,21 @@ Node* CmpPNode::Ideal(PhaseGVN *phase, bool can_reshape) {
     // input instead of the oop input to avoid keeping buffer allocations alive.
     return new CmpINode(in(1)->as_InlineType()->get_null_marker(), phase->intcon(0));
   }
+  if (in(1)->is_InlineType() || in(2)->is_InlineType()) {
+    // If one operand is an inline type, convert this to a "both operands are null" check:
+    //   CmpL(OrL(CastP2X(..), CastP2X(..)), 0L)
+    // CmpLNode::Ideal might optimize this further to avoid keeping buffer allocations alive.
```

Comment:
Why is this possible?

Reply:
As we discussed off-thread, this relies on CmpP being a buffer oop pointer comparison and since there are no guarantees that such pointers are equivalent, we can convert the check. I improved the comment and hopefully it's easier to follow now.